configuration/configurator/ConfiguratorAPI.cpp

```diff
@@ -0,0 +1,745 @@
+
+        //int index = CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getIndexOfElement(pElement);
+
+        //CElementArray *pElementArray = (static_cast<CElementArray*>(CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getComponent(index)->getParentNode()));
+
+        CElementArray *pElementArray = static_cast<CElementArray*>(pElement->getParentNode());
+
+        assert(pElementArray->getNodeType() == XSD_ELEMENT_ARRAY);
+
+        int nLength = pElementArray->length();
+
+        return nLength;
+    }
+}
+
+void* getComponentInConfiguration(int idx)
+{
+    assert(idx < CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getNumberOfComponents());
+
+    CElement *pElement = CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getComponent(idx);
+
+    return pElement;
+}
+
+void* getComponentInstance(int idx, void *pData)
+{
+    assert(pData != NULL);
+    assert(((static_cast<CElement*>(pData))->getNodeType()) == XSD_ELEMENT);
+
+    CElement *pElement = static_cast<CElement*>(pData);
+
+    CElementArray *pElementArray = static_cast<CElementArray*>(pElement->getParentNode());
+
+    //assert(pElementArray->length() > idx);
+
+    if (pElementArray->length() >= idx)
+        idx = 0;
+
+    CElement *pElement2 = &(pElementArray->item(idx));
+
+    return pElement2;
+}
+
+const char* getComponentNameInConfiguration(int idx, void *pData)
+{
+    if (pData == NULL)
+    {
+        return CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getComponent(idx)->getName();
+    }
+    else
+    {
+        CElement *pElement = static_cast<CElement*>(pData);
+
+        assert(pElement->getNodeType() == XSD_ELEMENT);
+
+        pElement = &(dynamic_cast<const CElementArray*>(pElement->getConstParentNode())->item(idx));
+
+        const CAttribute *pAttribute = NULL;
+
+        /*if (pElement->getAttributeArray() != NULL)
+        {
+            pAttribute = pElement->getAttributeArray()->findAttributeWithName("name");
+
+            return pAttribute->getName();
+        }
+        else
+        {*/
+            return NULL;
+        //}
+    }
+
+    assert(false);
+    return NULL;
+}
+
+
+const void* getPointerToComponentInConfiguration(int idx, void *pData, int compIdx)
+{
+    if (pData == NULL)
+    {
+        const CElement *pElement = CConfigSchemaHelper::getInstance()->getSchemaMapManager()->getComponent(idx);
+
+        assert(pElement != NULL);
+
+        const CXSDNodeBase *pNodeBase = pElement->getConstParentNode();
+        const CElementArray *pElementArray = dynamic_cast<const CElementArray*>(pNodeBase);
+
+        assert(pElementArray != NULL);
+
+        return pElementArray;
+    }
+    else
+    {
+        assert( compIdx >= 0);
+
+        CElementArray *pElementArray = static_cast<CElementArray*>(pData);
+
+        assert(pElementArray->getNodeType() == XSD_ELEMENT_ARRAY);
+
+        const CXSDNodeBase *pNodeBase = &(pElementArray->item(compIdx+idx));
+
```

Comment:
Are you being paid by the line? Why so many blank lines? They make it a lot harder to review the code (GitHub is struggling with the size of the change as it is).